extension/amqp/listener_test.go

```diff
@@ -0,0 +1,174 @@
+// +build unit
+
+package amqp_test
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/hellofresh/goengine/driver/sql"
+	"github.com/hellofresh/goengine/extension/amqp"
+	goengineLogger "github.com/hellofresh/goengine/extension/logrus"
+	"github.com/hellofresh/goengine/mocks"
+	"github.com/sirupsen/logrus"
+	"github.com/sirupsen/logrus/hooks/test"
+	libamqp "github.com/streadway/amqp"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestListener_Listen(t *testing.T) {
+	t.Run("Listen, consume and stop", func(t *testing.T) {
+		ensure := require.New(t)
+		ctrl := gomock.NewController(t)
+		defer ctrl.Finish()
+
+		ctx, ctxCancel := context.WithTimeout(context.Background(), time.Second)
+		defer ctxCancel()
+
+		consumeCalls := 0
+		consume := func() (io.Closer, <-chan libamqp.Delivery, error) {
+			consumeCalls++
+			ch := make(chan libamqp.Delivery, 2)
+			ch <- libamqp.Delivery{
+				Body: []byte(`{"no": 1, "aggregate_id": "8150276e-34fe-49d9-aeae-a35af0040a4f"}`),
+			}
+			ch <- libamqp.Delivery{
+				Body: []byte(`{"no": 2, "aggregate_id": "8150276e-34fe-49d9-aeae-a35af0040a4f"}`),
+			}
+			return nil, ch, nil
+		}
+		triggerCalls := 0
+		trigger := func(ctx context.Context, notification *sql.ProjectionNotification) error {
+			triggerCalls++
+			switch triggerCalls {
+			case 1:
+				ensure.Equal(&sql.ProjectionNotification{No: 1, AggregateID: "8150276e-34fe-49d9-aeae-a35af0040a4f"}, notification)
+			case 2:
+				ensure.Equal(&sql.ProjectionNotification{No: 2, AggregateID: "8150276e-34fe-49d9-aeae-a35af0040a4f"}, notification)
+				ctxCancel()
+			default:
+				ensure.Fail("Only 2 calls to trigger where expected")
+			}
+			return nil
+		}
+		logger := mocks.NewLogger(ctrl)
+
+		listener, err := amqp.NewListener(consume, time.Millisecond, time.Millisecond, logger)
+		ensure.NoError(err)
+
+		err = listener.Listen(ctx, trigger)
+
+		ensure.Equal(context.Canceled, err)
+		ensure.Equal(1, consumeCalls)
+		ensure.Equal(2, triggerCalls)
+	})
+
+	t.Run("Reconnect with exponential back-off", func(t *testing.T) {
+		ensure := require.New(t)
+		ctrl := gomock.NewController(t)
+		defer ctrl.Finish()
+
+		ctx, ctxCancel := context.WithTimeout(context.Background(), time.Second)
+		defer ctxCancel()
+
+		var (
+			consumeCalls []time.Time
+		)
```

Comment:
```suggestion
		var consumeCalls []time.Time
```

We can have this just on one line.